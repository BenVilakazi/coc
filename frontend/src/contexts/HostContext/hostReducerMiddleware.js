import socketInstance from '../../socket/socket';
import shuffleArray from '../../helpers/shuffleArray';

function closeGame() {
  socketInstance.closeSocket();
}

function createLobby() {
  socketInstance.createLobby();
}

function sendPlayerConnectedMessages(payload) {
  socketInstance.sendMessage({
    event: 'update',
    recipients: payload.players,
    payload: {
      gameState: 'connected',
      message: payload.message,
      removeLoading: 'joining-lobby',
    },
  });
}

function sendNameTakenMessage(payload) {
  socketInstance.sendMessage({
    event: 'update',
    recipients: [payload.playerId],
    payload: {
      gameState: 'connection-refused-name-taken',
      removeLoading: 'joining-lobby',
    },
  });
}

function playerConnected(payload) {
  socketInstance.sendMessage({
    event: 'update',
    recipients: [payload.playerId],
    payload: {
      gameState: 'connected',
      message: {
        big: 'Attempting to join lobby',
        small: 'Please wait',
      },
    },
  });
}

function sendCardsSubmittedMessage(payload) {
  socketInstance.sendMessage({
    event: 'update',
    recipients: [payload.playerId],
    payload: {
      gameState: 'cards-submitted',
      message: {
        big: 'WAIT FOR OTHER PLAYERS',
        small: 'Yell at them to hurry up if you wish',
      },
      removeLoading: 'submitting-cards',
    },
  });
}

function sendEndOfRoundMessages(payload) {
  const { losers, winnerName, winnerId, czar } = payload;

  socketInstance.sendMessage({
    recipients: [...losers],
    event: 'update',
    payload: {
      gameState: 'showing-end-round-messages',
      message: {
        big: `${winnerName} won this round`,
        small: 'Better luck next time, loser',
      },
    },
  });

  socketInstance.sendMessage({
    recipients: [winnerId],
    event: 'update',
    payload: {
      gameState: 'showing-end-round-messages',
      message: {
        big: 'Hey! You won!',
        small: 'Finally, now wait for the next round!',
      },
    },
  });

  socketInstance.sendMessage({
    recipients: [czar],
    event: 'update',
    payload: {
      gameState: 'showing-end-round-messages',
      message: {
        big: `You selected ${winnerName}`,
        small: "Now it's your turn to win",
      },
    },
  });
}

function sendKickPlayerMessage({ playerId }) {
  socketInstance.sendMessage({
    recipients: [playerId],
    event: 'update',
    payload: {
      gameState: 'player-kicked',
    },
  });

  socketInstance.sendMessage({
    event: 'kick-player',
    payload: { playerId },
  });
}

async function getDeck({ selectedPacks }) {
  const apiURL = process.env.REACT_APP_API_URL;
  const queryString = selectedPacks.join(',');
  const query = `${apiURL}/deck/cards?packs=${queryString}`;
  const errorMessage = `Error fetching cards. Query: ${query}`;
  const cardsRequest = await fetch(query);
  if (cardsRequest.ok) {
    const cards = await cardsRequest.json();
    if (cards.white.length) {
      return cards;
    }
  }
  throw new Error(errorMessage);
}

function sendCardsToPlayers({ selectedBlackCard, players, playerIDs }) {
  playerIDs.forEach((playerID) => {
    if (!players[playerID].isCzar) {
      socketInstance.sendMessage({
        event: 'deal-white-cards',
        payload: {
          cards: players[playerID].cards.map((card) => card.text),
          selectCardCount: selectedBlackCard.pick,
        },
        recipients: [playerID],
      });
    }
  });
}

function sendShuffleJoinCodeMessage() {
  socketInstance.sendMessage({
    event: 'shuffle-join-code',
    payload: {},
  });
}

function notifyCzar({ players, playerIDs }) {
  const czar = playerIDs.find((player) => players[player].isCzar);

  if (czar) {
    socketInstance.sendMessage({
      event: 'update',
      payload: {
        gameState: 'waiting-for-player-card-submissions',
        message: {
          big: "You're the Czar",
          small: 'Wait for the players to select their cards',
        },
      },
      recipients: [czar],
    });
  } else {
    throw new Error('Czar not found!');
  }
}

function czarSelectWinner({ players, playerIDs }) {
  // identify czar/not czar
  const czar = playerIDs.find((playerID) => players[playerID].isCzar);
  const notCzars = playerIDs.filter(
    (playerID) =>
      !players[playerID].isCzar &&
      (players[playerID].status === 'playing' ||
        (players[playerID].status === 'staging' &&
          players[playerID].hasSubmittedCards)),
  );

  // gather all players submitted cards
  const submittedCardsOrdered = notCzars.map((playerID) => ({
    playerID,
    cards: players[playerID].submittedCards.map(
      (card) => players[playerID].cards[card].text,
    ),
  }));

  // shuffle the cards
  const submittedCardsShuffled = shuffleArray(submittedCardsOrdered);

  socketInstance.sendMessage({
    event: 'update',
    payload: {
      gameState: 'select-winner',
      submittedCards: submittedCardsShuffled,
      selectCardCount: 1,
    },
    recipients: [czar],
  });

  socketInstance.sendMessage({
    event: 'update',
    payload: {
      gameState: 'waiting-for-czar',
      message: {
        big: 'the czar is selecting',
        small: 'For best results, watch the host screen',
      },
    },
    recipients: notCzars,
  });
}

function sendTooManyPlayersMessage(payload) {
  socketInstance.sendMessage({
    recipients: payload.players,
    event: 'update',
    payload: {
      gameState: 'error',
      message: {
        big: 'Player limit has been reached',
        small: '',
      },
      removeLoading: 'joining-lobby',
    },
  });

  payload.players.forEach((playerId) => {
    socketInstance.sendMessage({
      event: 'kick-player',
      payload: { playerId },
    });
  });
}

function sendEndOfGameMessages({ gameWinner, playerIDs }) {
  const losers = playerIDs.filter((playerID) => playerID !== gameWinner);

  socketInstance.sendMessage({
    event: 'update',
    payload: {
      gameState: 'end-game',
      message: {
        big: 'Congrats! You won it all!!1!',
        small: 'Not necessarily something to be proud of',
      },
    },
    recipients: [gameWinner],
  });

  if (losers.length) {
    socketInstance.sendMessage({
      event: 'update',
      payload: {
        gameState: 'end-game',
        message: {
          big: 'Loser ðŸ‘Žï¸Ž',
          small: `Better to lose the game than your integrity`,
        },
      },
      recipients: losers,
    });
  }
}

function removeKickedPlayersCard({ czarId, playerId }) {
  socketInstance.sendMessage({
    event: 'remove-disconnected-players-card',
    payload: {
      playerId,
    },
    recipients: [czarId],
  });
}

export default async function hostReducerMiddleware(
  { type, payload },
  dispatch,
) {
  switch (type) {
    case `CLOSE_GAME`:
      closeGame();
      break;

    case 'CREATE_LOBBY':
      createLobby();
      break;

    case 'PLAYER_RECONNECTED':
    case 'PLAYER_CONNECTED':
      playerConnected(payload);
      break;

    case 'SEND_PLAYER_CONNECTED_MESSAGES':
      return sendPlayerConnectedMessages(payload);

    case 'SEND_NAME_TAKEN_MESSAGE':
      return sendNameTakenMessage(payload);

    case 'PLAYER_SUBMIT':
      sendCardsSubmittedMessage(payload);
      break;

    case 'KICK_PLAYER':
      removeKickedPlayersCard(payload);
      sendKickPlayerMessage(payload);
      break;

    case 'SET_DECK': {
      try {
        const deck = await getDeck(payload);
        return dispatch({
          type: 'SET_DECK',
          payload: { deck },
        });
      } catch (err) {
        throw new Error(err);
      }
    }

    case 'SEND_CARDS_TO_PLAYERS':
      sendCardsToPlayers(payload);
      break;

    case 'SHUFFLE_JOIN_CODE':
      sendShuffleJoinCodeMessage();
      break;

    case 'NOTIFY_CZAR':
      notifyCzar(payload);
      break;

    case 'SKIP_UNSUBMITTED_PLAYERS':
    case 'CZAR_SELECT_WINNER':
      czarSelectWinner(payload);
      break;

    case 'SEND_END_OF_ROUND_MESSAGES':
      sendEndOfRoundMessages(payload);
      break;

    case 'TOO_MANY_PLAYERS':
      sendTooManyPlayersMessage(payload);
      break;

    case 'GAME_OVER':
      sendEndOfGameMessages(payload);
      break;

    default:
      break;
  }
  return dispatch({ type, payload });
}
