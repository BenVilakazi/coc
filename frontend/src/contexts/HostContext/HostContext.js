import React, { createContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import useReducerMiddleware from '../useReducerMiddleware';
import HostReducer from './HostReducer';
import socketInstance from '../../socket/socket';
import hostReducerMiddleware from './hostReducerMiddleware';
import config from '../../config';
import AlertModal from '../../components/Modal/AlertModal';

const propTypes = {
  children: PropTypes.node.isRequired,
};

const { emitter } = socketInstance;

const initialState = {
  gameState: 'waiting-for-lobby',
  lobbyID: '',
  players: {},
  playerIDs: [],
  gameSettings: {
    maxPlayers: config.maxPlayers.default,
    winningScore: config.winningScore.default,
    selectedPacks: config.initialSelectedPack,
    handSize: config.handSize,
    winnerScreenDisplayTime: config.winnerScreenDisplayTime,
    hideJoinCode: false,
  },
  deck: { black: [], white: [] },
  loading: [],
  czarSelection: '',
};

export const HostContext = createContext();

function HostProvider({ children }) {
  const [state, dispatch] = useReducerMiddleware(
    hostReducerMiddleware,
    HostReducer,
    initialState,
  );

  const [socketHasError, setSocketHasError] = useState(false);

  function createMessageHandler(currentState) {
    const czarId =
      currentState.playerIDs.length > 0 &&
      currentState.playerIDs.find(
        (playerId) => currentState.players[playerId].isCzar,
      );

    return ({ event, payload, sender }) => {
      switch (event) {
        case 'player-connected': {
          const allPlayers = [...Object.values(currentState.players)];
          const existingPlayer = allPlayers.find(
            (player) =>
              player.name.toUpperCase() === payload.playerName.toUpperCase(),
          );

          if (!existingPlayer) {
            return dispatch({ type: 'PLAYER_CONNECTED', payload });
          }

          if (existingPlayer.status === 'disconnected') {
            return dispatch({ type: 'PLAYER_RECONNECTED', payload });
          }

          return dispatch({ type: 'SEND_NAME_TAKEN_MESSAGE', payload });
        }
        case 'player-disconnected':
          return dispatch({
            type: 'PLAYER_DISCONNECTED',
            payload: {
              ...payload,
              czarId,
            },
          });

        case 'preview-winner':
          return dispatch({ type: 'PREVIEW_WINNER', payload });

        case 'winner-selected':
          return dispatch({ type: 'WINNER_SELECTED', payload });

        case 'lobby-created':
          return dispatch({ type: 'SET_LOBBY_ID', payload });

        case 'black-card-selected':
          return dispatch({ type: 'SELECT_BLACK_CARD', payload });

        case 'join-code-shuffled':
          return dispatch({ type: 'UPDATE_JOIN_CODE', payload });

        case 'player-submit':
          return dispatch({
            type: 'PLAYER_SUBMIT',
            payload: { ...payload, playerId: sender },
          });

        case 'socket-connection-error':
          dispatch({ type: 'UPDATE_JOIN_CODE', payload: { lobbyID: 'ERROR' } });
          return setSocketHasError(true);

        default:
          return undefined;
      }
    };
  }

  useEffect(() => {
    const handleMessage = createMessageHandler(state);
    emitter.on('message', handleMessage);
    return () => {
      emitter.off('message', handleMessage);
    };
  }, [state]);

  useEffect(() => {
    const stagedPlayerIDs = state.playerIDs.filter(
      (playerID) => state.players[playerID].status === 'staging',
    );

    const reconnectingPlayersIds = state.playerIDs.filter(
      (playerID) => state.players[playerID].status === 're-connected',
    );

    if (reconnectingPlayersIds.length > 0) {
      const reConnectingPlayers = reconnectingPlayersIds.reduce(
        (acc, id) => ({ ...acc, [id]: state.players[id] }),
        {},
      );

      dispatch({
        type: 'SEND_CARDS_TO_PLAYERS',
        payload: {
          selectedBlackCard: state.selectedBlackCard,
          players: reConnectingPlayers,
          playerIDs: reconnectingPlayersIds,
        },
      });
    }

    if (stagedPlayerIDs.length) {
      const numberOfPlayersExceedingLimit =
        state.playerIDs.length - state.gameSettings.maxPlayers;

      if (numberOfPlayersExceedingLimit > 0) {
        const playerIDsExceedingLimit = state.playerIDs.slice(
          -numberOfPlayersExceedingLimit,
        );

        dispatch({
          type: 'TOO_MANY_PLAYERS',
          payload: {
            players: playerIDsExceedingLimit,
          },
        });
      } else {
        const message =
          state.gameState === 'waiting-for-players'
            ? {
                big: "You've joined the lobby",
                small: 'Please wait for the host to start the game',
              }
            : {
                big: 'A round is in progress',
                small: 'You will join the next round automatically',
              };

        dispatch({
          type: 'SEND_PLAYER_CONNECTED_MESSAGES',
          payload: {
            players: stagedPlayerIDs,
            message,
          },
        });
      }
    }
  }, [state.playerIDs]);

  useEffect(() => {
    // if there is one player AND they are not waiting for the game to begin, end the game
    if (
      state.playerIDs.filter(
        (playerID) => state.players[playerID].status === 'playing',
      ).length === 1 &&
      state.gameState !== 'game-over'
    ) {
      dispatch({
        type: 'GAME_OVER',
        payload: { gameWinner: state.playerIDs[0], playerIDs: state.playerIDs },
      });
    }
  }, [state.playerIDs, state.players]);

  return (
    <HostContext.Provider value={{ state, dispatch }}>
      {children}
      {socketHasError && (
        <AlertModal
          bigText="SOCKET ERROR"
          smallText="Please try again later"
          buttonText="Click anywhere to restart"
          onClick={() => window.location.reload()}
        />
      )}
    </HostContext.Provider>
  );
}

HostProvider.propTypes = propTypes;

export default HostProvider;
