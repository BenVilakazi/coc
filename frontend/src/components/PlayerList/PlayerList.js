import React, { useContext } from 'react';
import styled from 'styled-components';
import { HostContext } from '../../contexts/HostContext/HostContext';
import BlackCard from '../../assets/black-card-icon.svg';
import OffsetWhiteCardStack from '../../assets/offset-white-card-stack-icon.svg';
import TallyCount from '../TallyMarker/TallyCount';

const PlayerTable = styled.div`
  display: flex;
  flex-direction: column;

  width: 486px;
  padding: 40px;
  background-color: var(--primary-background-color);

  position: relative;

  & > div:not(:last-child) {
    margin-bottom: 24px;
  }
`;

const PlayerRow = styled.div`
  display: flex;
  align-items: center;
  width: 100%;

  .player-info {
    display: flex;
    align-items: center;
    justify-content: space-between;

    flex: 1;
    margin: ${(props) => (props.isCzar ? '0 12px 0 26px' : '0 24px 0 26px')};
    padding: ${(props) => (props.isCzar ? '7px 12px' : '7px 0')};
    border-bottom: ${(props) => (props.isCzar ? 0 : '1px solid #ccc')};
    border-radius: ${(props) => (props.isCzar ? '4px' : 0)};

    background-color: ${(props) =>
      props.isCzar
        ? 'var(--secondary-background-color);'
        : 'var(--primary-background-color)'};

    h1 {
      text-transform: uppercase;
      font-size: 24px;
      font-weight: 900;
      letter-spacing: 1px;
      line-height: 1.5em;
      color: ${(props) => {
        if (props.isCzar) {
          return 'var(--secondary-text-color)';
        }
        if (props.faded) {
          return 'var(--accent-text-color)';
        }
        return 'var(--primary-text-color)';
      }};
      text-decoration: ${(props) =>
        props.isDisconnected ? 'line-through' : 'none'};
    }

    span {
      color: ${(props) =>
        props.isCzar
          ? 'var(--secondary-text-color)'
          : 'var(--primary-text-color)'};
      font-size: 24px;
      font-weight: 900;
    }
  }
`;
function PlayerList() {
  const { state } = useContext(HostContext);
  const { playerIDs, players } = state;

  const playersArray = [...playerIDs.map((id) => players[id])];

  return (
    <PlayerTable data-testid="playerList-container">
      {playersArray.map((player) => {
        const showIcon = player.submittedCards.length > 0 || player.isCzar;
        const playerIcon = player.isCzar ? BlackCard : OffsetWhiteCardStack;
        const playerKey =
          playerIDs.find((id) => players[id] === player) || player.playerId;

        return (
          <PlayerRow
            key={playerKey}
            isCzar={player.isCzar}
            faded={
              player.status === 'staging' || player.status === 'disconnected'
            }
            isDisconnected={player.status === 'disconnected'}
            data-testid={`row-${player.name}`}
          >
            <img
              src={playerIcon}
              alt="card icon"
              style={{ visibility: showIcon ? 'visible' : 'hidden' }}
              data-testid={`icon-${player.name}`}
            />

            <div className="player-info">
              <h1>{player.name}</h1>
              <TallyCount
                score={player.score}
                color={player.isCzar ? 'secondary' : 'primary'}
                faded={
                  player.status === 'staging' ||
                  player.status === 'disconnected'
                }
              />
            </div>
          </PlayerRow>
        );
      })}
    </PlayerTable>
  );
}

export default PlayerList;
