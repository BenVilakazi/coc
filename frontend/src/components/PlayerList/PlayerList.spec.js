import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { HostContext } from '../../contexts/HostContext/HostContext';

import PlayerList from './PlayerList';

describe('PlayerList', () => {
  it('renders PlayerList component given a simple playerList object', () => {
    const state = {
      players: {
        playerID1: {
          name: 'Foo',
          score: 3,
          isCzar: false,
          submittedCards: [1, 4],
          cards: [
            { text: 'test', pack: 0 },
            { text: 'test', pack: 0 },
          ],
          status: 'playing',
        },
        playerID2: {
          name: 'Bar',
          score: 5,
          isCzar: true,
          submittedCards: [],
          cards: [],
          status: 'playing',
        },
      },
      playerIDs: ['playerID1', 'playerID2'],
    };

    render(
      <HostContext.Provider value={{ state }}>
        <PlayerList />
      </HostContext.Provider>,
    );

    expect(screen.getByText(state.players.playerID1.name)).toBeInTheDocument();
    expect(screen.getByText(state.players.playerID2.name)).toBeInTheDocument();
  });

  it('renders players with status of "playing" and "disconnected"', () => {
    const state = {
      players: {
        playerID1: {
          name: 'Foo',
          score: 3,
          isCzar: false,
          submittedCards: [1, 4],
          cards: [
            { text: 'test', pack: 0 },
            { text: 'test', pack: 0 },
          ],
          status: 'playing',
        },
        playerID2: {
          name: 'Bar',
          score: 5,
          isCzar: true,
          submittedCards: [],
          cards: [],
          status: 'staging',
        },
        playerID3: {
          name: 'Baz',
          score: 5,
          isCzar: true,
          submittedCards: [],
          cards: [],
          status: 'disconnected',
        },
      },
      playerIDs: ['playerID1', 'playerID2', 'playerID3'],
    };

    render(
      <HostContext.Provider value={{ state }}>
        <PlayerList />
      </HostContext.Provider>,
    );

    expect(
      screen.getByText(state.players[state.playerIDs[0]].name),
    ).toBeInTheDocument();
    expect(
      screen.getByText(state.players[state.playerIDs[1]].name),
    ).toBeInTheDocument();
    expect(
      screen.getByText(state.players[state.playerIDs[2]].name),
    ).toBeInTheDocument();
  });

  it('renders players with all possible statuses with correct stylings', () => {
    const state = {
      players: {
        playerID1: {
          name: 'Foo',
          score: 3,
          isCzar: false,
          submittedCards: [1, 4],
          cards: [
            { text: 'test', pack: 0 },
            { text: 'test', pack: 0 },
          ],
          status: 'playing',
        },
        playerID2: {
          name: 'Bar',
          score: 5,
          isCzar: true,
          submittedCards: [],
          cards: [],
          status: 'staging',
        },
        playerID3: {
          name: 'Baz',
          score: 5,
          isCzar: true,
          submittedCards: [],
          cards: [],
          status: 'disconnected',
        },
      },
      playerIDs: ['playerID1', 'playerID2', 'playerID3'],
    };

    const tree = renderer
      .create(
        <HostContext.Provider value={{ state }}>
          <PlayerList />
        </HostContext.Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('If player has submitted his cards or player is czar, render the icon fully visible', () => {
    const state = {
      players: {
        playerID1: {
          name: 'Foo',
          score: 3,
          isCzar: false,
          submittedCards: [1, 4],
          cards: [
            { text: 'test', pack: 0 },
            { text: 'test', pack: 0 },
          ],
        },
        playerID2: {
          name: 'Bar',
          score: 5,
          isCzar: true,
          submittedCards: [],
          cards: [],
        },
      },
      playerIDs: ['playerID1', 'playerID2'],
    };

    render(
      <HostContext.Provider value={{ state }}>
        <PlayerList />
      </HostContext.Provider>,
    );

    expect(
      screen.getByTestId(`icon-${state.players.playerID1.name}`),
    ).toBeVisible();
    expect(
      screen.getByTestId(`icon-${state.players.playerID2.name}`),
    ).toBeVisible();
  });

  it('If player has NOT submitted his cards, render the icon with visibility hidden', () => {
    const state = {
      players: {
        playerID1: {
          name: 'Foo',
          score: 3,
          isCzar: false,
          submittedCards: [],
          cards: [],
        },
      },
      playerIDs: ['playerID1'],
    };

    render(
      <HostContext.Provider value={{ state }}>
        <PlayerList />
      </HostContext.Provider>,
    );

    expect(
      screen.getByTestId(`icon-${state.players.playerID1.name}`),
    ).not.toBeVisible();
  });

  it('If playerIDs is empty, PlayerList still renders as an empty container', () => {
    const state = {
      players: {},
      playerIDs: [],
    };

    render(
      <HostContext.Provider value={{ state }}>
        <PlayerList />
      </HostContext.Provider>,
    );

    expect(screen.getByTestId('playerList-container')).toBeInTheDocument();
  });

  it('If the player object goes in a different order than playerIDs, the component follow playerIDs order', () => {
    const state = {
      players: {
        playerID1: {
          name: 'Foo',
          score: 3,
          isCzar: false,
          submittedCards: [1, 4],
          cards: [],
        },
        playerID2: {
          name: 'Bar',
          score: 5,
          isCzar: true,
          submittedCards: [],
          cards: [],
        },
      },
      playerIDs: ['playerID2', 'playerID1'],
    };

    render(
      <HostContext.Provider value={{ state }}>
        <PlayerList />
      </HostContext.Provider>,
    );

    const player2 = screen.getByTestId('row-Bar');

    expect(screen.getByTestId('playerList-container').firstChild).toBe(player2);
  });

  // TODO: test('If player is czar, render the its row with the proper colors', () => {})
});
