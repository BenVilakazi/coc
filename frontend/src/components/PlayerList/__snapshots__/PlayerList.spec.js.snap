// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerList renders players with all possible statuses with correct stylings 1`] = `
.c3 {
  stroke: var(--primary-text-color);
}

.c5 {
  stroke: var(--accent-text-color);
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 486px;
  padding: 40px;
  background-color: var(--primary-background-color);
  position: relative;
}

.c0 > div:not(:last-child) {
  margin-bottom: 24px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c1 .player-info {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0 24px 0 26px;
  padding: 7px 0;
  border-bottom: 1px solid #ccc;
  border-radius: 0;
  background-color: var(--primary-background-color);
}

.c1 .player-info h1 {
  text-transform: uppercase;
  font-size: 24px;
  font-weight: 900;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1.5em;
  color: var(--primary-text-color);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 .player-info span {
  color: var(--primary-text-color);
  font-size: 24px;
  font-weight: 900;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c4 .player-info {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0 12px 0 26px;
  padding: 7px 12px;
  border-bottom: 0;
  border-radius: 4px;
  background-color: var(--secondary-background-color);
}

.c4 .player-info h1 {
  text-transform: uppercase;
  font-size: 24px;
  font-weight: 900;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1.5em;
  color: var(--secondary-text-color);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 .player-info span {
  color: var(--secondary-text-color);
  font-size: 24px;
  font-weight: 900;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c6 .player-info {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0 12px 0 26px;
  padding: 7px 12px;
  border-bottom: 0;
  border-radius: 4px;
  background-color: var(--secondary-background-color);
}

.c6 .player-info h1 {
  text-transform: uppercase;
  font-size: 24px;
  font-weight: 900;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1.5em;
  color: var(--secondary-text-color);
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
}

.c6 .player-info span {
  color: var(--secondary-text-color);
  font-size: 24px;
  font-weight: 900;
}

<div
  className="c0"
  data-testid="playerList-container"
>
  <div
    className="c1"
    data-testid="row-Foo"
  >
    <img
      alt="card icon"
      data-testid="icon-Foo"
      src="offset-white-card-stack-icon.svg"
      style={
        Object {
          "visibility": "visible",
        }
      }
    />
    <div
      className="player-info"
    >
      <h1>
        Foo
      </h1>
      <div
        className="c2"
      >
        <svg
          data-testid="tally-svg"
          fill="none"
          height="30"
          viewBox="0 0 42 30"
          width="42"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              className="c3"
              color="primary"
              d="M32.4285 3.19775V29.5701"
              data-testid="tally"
              strokeWidth="2"
            />
            <path
              className="c3"
              color="primary"
              d="M25.4439 3.51099L23.2217 29.5966"
              data-testid="tally"
              strokeWidth="2"
            />
            <path
              className="c3"
              color="primary"
              d="M16.5548 3.51099L14.3326 29.5966"
              data-testid="tally"
              strokeWidth="2"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>
  <div
    className="c4"
    data-testid="row-Bar"
  >
    <img
      alt="card icon"
      data-testid="icon-Bar"
      src="black-card-icon.svg"
      style={
        Object {
          "visibility": "visible",
        }
      }
    />
    <div
      className="player-info"
    >
      <h1>
        Bar
      </h1>
      <div
        className="c2"
      >
        <svg
          data-testid="tally-svg"
          fill="none"
          height="30"
          viewBox="0 0 42 30"
          width="42"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              className="c5"
              color="secondary"
              d="M32.4285 3.19775V29.5701"
              data-testid="tally"
              strokeWidth="2"
            />
            <path
              className="c5"
              color="secondary"
              d="M25.4439 3.51099L23.2217 29.5966"
              data-testid="tally"
              strokeWidth="2"
            />
            <path
              className="c5"
              color="secondary"
              d="M16.5548 3.51099L14.3326 29.5966"
              data-testid="tally"
              strokeWidth="2"
            />
            <path
              className="c5"
              color="secondary"
              d="M7.66589 3.51099L5.44367 29.5966"
              data-testid="tally"
              strokeWidth="2"
            />
            <path
              className="c5"
              color="secondary"
              d="M0.916168 0.599211L40.9161 29.1692"
              data-testid="tally"
              strokeWidth="2"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>
  <div
    className="c6"
    data-testid="row-Baz"
  >
    <img
      alt="card icon"
      data-testid="icon-Baz"
      src="black-card-icon.svg"
      style={
        Object {
          "visibility": "visible",
        }
      }
    />
    <div
      className="player-info"
    >
      <h1>
        Baz
      </h1>
      <div
        className="c2"
      >
        <svg
          data-testid="tally-svg"
          fill="none"
          height="30"
          viewBox="0 0 42 30"
          width="42"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              className="c5"
              color="secondary"
              d="M32.4285 3.19775V29.5701"
              data-testid="tally"
              strokeWidth="2"
            />
            <path
              className="c5"
              color="secondary"
              d="M25.4439 3.51099L23.2217 29.5966"
              data-testid="tally"
              strokeWidth="2"
            />
            <path
              className="c5"
              color="secondary"
              d="M16.5548 3.51099L14.3326 29.5966"
              data-testid="tally"
              strokeWidth="2"
            />
            <path
              className="c5"
              color="secondary"
              d="M7.66589 3.51099L5.44367 29.5966"
              data-testid="tally"
              strokeWidth="2"
            />
            <path
              className="c5"
              color="secondary"
              d="M0.916168 0.599211L40.9161 29.1692"
              data-testid="tally"
              strokeWidth="2"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
